### ISSUE TITLE
Lasership

### ISSUE THREAD

**Issue by adgaudio** [jkeen/tracking_number_data#4]
I found these tracking number formats for lasership:

Standard one:
10 chars:    LX16641251

And others:
15 chars:        1LS713461643483
18 chars:      1LS7235012355811-1
20 chars:    1LS72301001401854622
21 chars:   1LS716317038000783263

and this documentation:
https://www.trackingex.com/lasership-tracking.html

TODO: figure out if there is a matching checkdigit algo for each pattern and add support for this.

**Comment by adgaudio** [comment]
Waiting to acquire a larger dataset of lasership tracking numbers before comfortable submitting PR for this...

**Comment by adgaudio** [comment]
I started geeking out a little bit last hour or so on our lasership dataset, to figure out what the check digit algo is.  I didn't find one, but sharing some interesting info.  As much for me for later reference as anyone else :)

Looking at around 1500 1LS7 15 digit barcodes...

```
digit      0  1  2  3  4  5  6  7  8   9  10  11  12  13  14
unique     1  1  1  1  2  8  7  8  7  10  10  10  10  10  10
topval     1  L  S  7  2  4  5  0  0   5   1   2   7   5   9
```

On 15 digit 1LS7 numbers, the digits at index 4:9 look like zip codes but aren't.  (I cross referenced with a list of known zipcodes.  Several weren't found, and of zipcodes that were "valid" they were from kind of random places like Sherrill, NY (least populated town in NY)).  Perhaps these numbers are a LaserShip shipping center location, though.  

The digits at index 9:  look like they are incremental with time, because I correlated them with an incremental ID in our system that roughly matches passing of time.  However, digit 9 itself might be randomly chosen. The pic below shows digits 9 through 15 of a 15 digit 1LS7 barcode.  This is a surprising result.  The second pic below shows digits 10 through 15.  Slightly less surprising, but the correlation with time still suggests some kind of counter exists.
![image](https://user-images.githubusercontent.com/860369/36232069-58c3d5c8-11ae-11e8-885e-33126022af67.png)
![image](https://user-images.githubusercontent.com/860369/36232185-c53b97d6-11ae-11e8-904a-edee1a5a941a.png)

I dug a tiny little bit further looking at distribution of each digit for each position in the barcode.  Y axis is the raw digit.   X axis is the index in the barcode string.  Color is the number of times that digit appeared.
First picture is the heatmap for all digits in barcode.  Second picture is the heatmap for digits 9 to 15.
![image](https://user-images.githubusercontent.com/860369/36232476-e8fe5540-11af-11e8-9b6e-1a136beb7b5f.png)
![image](https://user-images.githubusercontent.com/860369/36232511-0a2d5950-11b0-11e8-803c-6bf7e04263a1.png)

The heatmap shows that digits 10 to end are pretty random, so possible lasership is in fact using some sort of counter.  My guess is they randomly choose the 9th digit and have "counters" running one per digit.  

If anyone knows how to reverse engineer hash functions, I'd love to learn how to figure out the function that generates lasership tracking numbers.  I guess next step would be to brute force a bunch of check functions (I tried against the ones we have in our repo once).

---

Anyways adding another new LS format with no check digit algo shortly (or next week).

**Comment by jkeen** [comment]
This is super cool Alex! I'm thinking the silliest/simplest answer is probably the correct one; that they've just got a counter. I sent them a tweet, let's see if they respond.

**Comment by jkeen** [comment]
I just rebased the `lasership` branch to reflect the changes in `master` (tests and things). I think we can delete the `dev` branch, can't we? The only thing that's different from the released and in use `master` branch is the lasership work which has stagnated a bit (no response from lasership, mysteries about check digits, etc). So `dev` is basically a lasership feature branch, but there's other changes to be made/released before we resolve that.

I think clients should pin their submodules to a specific commit (or a tag) off master, and then feature work should happen on feature branches and get PR'd into master. Not much need for dev then, right?

**Comment by TheBestMoshe** [comment]
Any updates on merging this into the main branch. In order to track Amazon deliveries, we need support for Lasership.

If it helps, I can try to build a dataset of Lasership tracking codes.

**Comment by jkeen** [comment]
Merged in #62
